pylint/checkers/typecheck.py

```diff
@@ -1900,13 +1900,34 @@ def _detect_unsupported_alternative_union_syntax(self, node: nodes.BinOp) -> Non
                 self._check_unsupported_alternative_union_syntax(node)
 
     def _check_unsupported_alternative_union_syntax(self, node: nodes.BinOp) -> None:
-        """Check if left or right node is of type `type`."""
+        """Check if left or right node is of type `type`.
+
+        If either is, and doesn't
+        support an or operator via a metaclass, infer that this is a mistaken
+        attempt to use alternative union syntax when not supported.
+        """
         msg = "unsupported operand type(s) for |"
-        for n in (node.left, node.right):
-            n = astroid.helpers.object_type(n)
-            if isinstance(n, nodes.ClassDef) and is_classdef_type(n):
-                self.add_message("unsupported-binary-operation", args=msg, node=node)
-                break
+        left_obj = astroid.helpers.object_type(node.left)
+        right_obj = astroid.helpers.object_type(node.right)
+        left_is_type = False
+        right_is_type = False
+
+        if isinstance(left_obj, nodes.ClassDef) and is_classdef_type(left_obj):
+            try:
+                left_obj.getattr("__or__")
+                return
+            except astroid.NotFoundError:
+                left_is_type = True
```

Comment:
If you mean Python built-in `hasattr` then I don't think so, because we need to look up the attribute in Astroid's inferred knowledge of the real class, not in the `left_obj` runtime object. If there's a Pylint equivalent to `hasattr` on inferred representations then we could use that, but I couldn't find one.

Reply:
Do you think it's worth it to add hasattr in astroid ?

Reply:
Sure, I'd be happy to give it a go.